notion-server/src/ethereum/mod.rs

```diff
@@ -1,25 +1,122 @@
+#![allow(unused)]
+
+use async_std::sync::{channel, Receiver, Sender};
+
+use snafu::{OptionExt, Snafu};
+
 use std::collections::HashMap;
 
+/// Shorthand for result types returned from the Simulation simulation.
+pub type Result<V, E = Error> = std::result::Result<V, E>;
+
+/// Errors arising from the simulation.
+#[derive(Debug, Snafu)]
+pub enum Error {
+    /// Operation was cancelled because the simulation is shutting down.
+    Terminated,
+}
+
+#[derive(Debug)]
+enum Operation {
+    CreateExecutionEnvironment(args::CreateExecutionEnvironment, Sender<EeIndex>),
+    CreateShardChain(args::CreateShardChain, Sender<u32>),
+    CreateShardBlock(args::CreateShardBlock, Sender<u32>),
+    GetShardBlock(args::GetShardBlock, Sender<ShardBlock>),
+}
+
+#[derive(Debug, Clone)]
+pub struct Handle(Sender<Operation>);
+
+impl Handle {
+    pub async fn create_shard_chain(&self, arg: args::CreateShardChain) -> Result<u32> {
+        let (sender, receiver) = channel(1);
+
+        self.0.send(Operation::CreateShardChain(arg, sender)).await;
+
+        receiver.recv().await.context(Terminated)
+    }
+
+    pub async fn create_shard_block(&self, arg: args::CreateShardBlock) -> Result<u32> {
+        let (sender, receiver) = channel(1);
+
+        self.0.send(Operation::CreateShardBlock(arg, sender)).await;
+
+        receiver.recv().await.context(Terminated)
+    }
+
+    pub async fn create_execution_environment(
+        &self,
+        arg: args::CreateExecutionEnvironment,
+    ) -> Result<EeIndex> {
+        let (sender, receiver) = channel(1);
+
+        self.0
+            .send(Operation::CreateExecutionEnvironment(arg, sender))
+            .await;
+
+        receiver.recv().await.context(Terminated)
+    }
+
+    pub async fn shard_block(&self, arg: args::GetShardBlock) -> Result<ShardBlock> {
+        let (sender, receiver) = channel(1);
+
+        self.0.send(Operation::GetShardBlock(arg, sender)).await;
+
+        receiver.recv().await.context(Terminated)
+    }
+}
+
 #[derive(Debug)]
-pub struct EthereumSimulation {
```

Comment:
IMO the slightly more verbose name is better because it's more clear, but I don't feel *too* strongly here (and I realize this file is already in an `ethereum` folder.

Reply:
Just throwing this out there -- Rust [RFC 356](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#avoid-redundant-prefixes-[rfc-356]) prefers that module items are not prefixed by the name of the module.

Reply:
Yep, makes sense, `Go` also has conventions against what they call "stuttering".  Cool, I think you guys are in the right here, let's call it `Simulation`.